lib/hashset.rs

```diff
@@ -0,0 +1,387 @@
+    }
+}
+
+// Set update semantics: update contains values that are in one of the sets
+// but not the other.
+impl<T: FromRecord + Hash + Eq + Clone + Ord> Mutator<HashSet<T>> for Record {
+    fn mutate(&self, set: &mut HashSet<T>) -> StdResult<(), String> {
+        let upd = <HashSet<T>>::from_record(self)?;
+        for v in upd.into_iter() {
+            if set.remove(&v).is_none() {
+                set.insert(v);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<T: Hash + Eq + Debug + Ord> Debug for HashSet<T> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        f.debug_set().entries(self.iter()).finish()
+    }
+}
+
+pub fn hashset_size<T>(s: &HashSet<T>) -> std_usize {
+    s.len() as std_usize
+}
+
+pub fn hashset_empty<T>() -> HashSet<T> {
+    HashSet::new()
+}
+
+pub fn hashset_singleton<T: Hash + Eq + Clone>(v: &T) -> HashSet<T> {
+    HashSet::unit((*v).clone())
+}
+
+pub fn hashset_insert<T: Hash + Eq + Clone>(s: &mut HashSet<T>, v: &T) {
+    s.insert(v.clone());
+}
+
+pub fn hashset_insert_imm<T: Hash + Eq + Clone>(s: &HashSet<T>, v: &T) -> HashSet<T> {
+    s.update((*v).clone())
+}
+
+pub fn hashset_remove<T: Hash + Eq + Clone>(s: &mut HashSet<T>, v: &T) {
+    s.remove(v);
+}
+
+pub fn hashset_remove_imm<T: Hash + Eq + Clone>(s: &HashSet<T>, v: &T) -> HashSet<T> {
+    s.without(v)
+}
+
+pub fn hashset_contains<T: Hash + Eq>(s: &HashSet<T>, v: &T) -> bool {
+    s.contains(v)
+}
+
+pub fn hashset_is_empty<T>(s: &HashSet<T>) -> bool {
+    s.is_empty()
+}
+
+pub fn hashset_nth<T: Hash + Clone + Ord>(s: &HashSet<T>, n: &std_usize) -> DDOption<T> {
+    option2std(s.iter().nth(*n as usize).cloned())
+}
+
+pub fn hashset_to_vec<T: Hash + Clone + Ord>(s: &HashSet<T>) -> DDVec<T> {
+    DDVec {
+        vec: s.iter().cloned().collect(),
+    }
+}
+
+pub fn hashset_union<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
+    s1.clone().union(s2.clone())
+}
+
+pub fn hashset_unions<T: Hash + Eq + Clone + Ord>(sets: &Vec<HashSet<T>>) -> HashSet<T> {
+    HashSet::unions(sets.iter().cloned())
+}
+
+pub fn group_hashset_unions<K, T: Hash + Eq + Clone + Ord>(
+    sets: &Group<K, HashSet<T>>,
+) -> HashSet<T> {
+    HashSet::unions(sets.iter().map(|tuple2(x, _)| x))
+}
+
+pub fn hashset_intersection<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
+    s1.clone().intersection(s2.clone())
+}
+
+pub fn hashset_difference<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
+    s1.clone().difference(s2.clone())
+}
+
+pub fn hashset_arg_min<T: Hash + Clone + Ord, B: Ord>(
+    s: &HashSet<T>,
+    f: &Box<dyn Closure<*const T, B>>,
+) -> DDOption<T> {
+    DDOption::from(s.iter().min_by_key(|x| f.call(*x)).map(|x| (*x).clone()))
+}
+
+pub fn hashset_arg_max<T: Hash + Clone + Ord, B: Ord>(
+    s: &HashSet<T>,
+    f: &Box<dyn Closure<*const T, B>>,
+) -> DDOption<T> {
+    DDOption::from(s.iter().max_by_key(|x| f.call(*x)).map(|x| (*x).clone()))
```

Comment:
By the looks of `Closure<*const T, _>` `Closure::call()` should be an unsafe function